### PR TITLE
basic CI workflow + Make compatible with python 3.8

### PR DESCRIPTION
**Pull request by orm011**
If it doesn't immediately work for you, 
you may need to manually enable the workflows in some setting for this to work.

You can ignore the conda stuff, current workflow for installing using poetry should work.
I added bc of my own mac setup.

- - -
This change is [Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/46)

### CODE REVIEW

**Review by orm011 — COMMENTED** [review]
Moved stuff to makefile, currently no need for conda (bc probably better to include it in our .whl, which will be ready soon.). 

> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/46#-:-NkwFuSyE8PlivfqvKkG:b5ez4z)* status: 0 of 10 files reviewed, 6 unresolved discussions (waiting on @mkornacker)

___
*[`.github/workflows/install_and_test.yml` line 1 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/46#-NkX1mZd6ACfnQaFOrGS:-NkwF6Wq2w36K8UcWw-Z:b-aprzdx) ([raw file](https://github.com/mkornacker/pixeltable/blob/a7de85d7d8a701a8205558b72d67eeb8b9b2ff0e/.github/workflows/install_and_test.yml#L1)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

add short explanatory comments to each step
</blockquote></details>

done.  there's currently no need for conda (docker is used in CI). this some of the script. That and the makefile, where I've refactored some of the steps.
___
*[`pixeltable/metadata/schema.py` line 104 at r2](https://reviewable.io/reviews/mkornacker/pixeltable/46#-NkwFYVp9_3D-UoxikHg:-NkwFYVp9_3D-UoxikHh:bsqhe6h) ([raw file](https://github.com/mkornacker/pixeltable/blob/9cdeb1eaccd619c11936f000d9da5e9e3a9a1088/pixeltable/metadata/schema.py#L104)):*
> ```Python
> 
>     # filter predicate applied to the base table; view-only
>     predicate: Optional[dict]
> ```

Needed to do this,  breaks tests in python 3.8. (ok in later)
___
*[`pixeltable/tests/test_functions.py` line 16 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/46#-NkX58QJ0HKWkHImEAdi:-NkwFix29LGTcUKeY2rJ:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/a7de85d7d8a701a8205558b72d67eeb8b9b2ff0e/pixeltable/tests/test_functions.py#L16)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

instead of disabling, comment out the failing assert and leave a TODO above it
</blockquote></details>

Done.
___
*[`conda_environment.yml` line 5 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/46#-NkX5U73DqMJ1oL3n-BI:-NkwFoKc4mk0IdCALT_m:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/a7de85d7d8a701a8205558b72d67eeb8b9b2ff0e/conda_environment.yml#L5)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

3.8?
</blockquote></details>

Done.
___
*[`dev_install.bash` line 9 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/46#-NkX5rmh0PF7A6s7thr7:-NkwFpY26mmzDi5r09hQ:b-j6e1vg) ([raw file](https://github.com/mkornacker/pixeltable/blob/a7de85d7d8a701a8205558b72d67eeb8b9b2ff0e/dev_install.bash#L9)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

'pwd' as default?
</blockquote></details>

done, in makefile.
___
*[`dev_install.bash` line 16 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/46#-NkX6MaV9GrqZPwsw2UO:-NkwFsQn82uc1s4hYdOg:bpyxtex) ([raw file](https://github.com/mkornacker/pixeltable/blob/a7de85d7d8a701a8205558b72d67eeb8b9b2ff0e/dev_install.bash#L16)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

"for running pytest --nbmake"
</blockquote></details>

makefile now.

File: .github/workflows/install_and_test.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+name: Install and test dev version
```
Comment (mkornacker) [inline .github/workflows/install_and_test.yml]: Call this something like "Main CI"?

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	@poetry run pytest --overwrite --nbmake --nbmake-kernel=$(KERNEL_NAME) docs/source/tutorials/*.ipynb
+
+.PHONY: typecheck
+typecheck: install
+	@echo "Running type checks..."
+	@poetry run mypy --config-file mypy.ini --package pixeltable
+
+.PHONY: clean
+clean:
+	@rm -f *.mp4 docs/source/tutorials/*.mp4 || true
```
Comment (mkornacker) [inline Makefile:50]: what's that for?

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+poetry.lock: pyproject.toml
+	@poetry lock --no-update
+
+# use a file to track whether the install has been run
+# avoid re-running the install if install has been done and deps 
+# haven't changed
+.make-install: poetry.lock
+	@echo "Installing development environment..."
+	@poetry install --all-extras --with=test
```
Comment (mkornacker) [inline Makefile]: --with=test,docs

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/46)* status: 0 of 11 files reviewed, 9 unresolved discussions (waiting on @mkornacker)

___
*[`Makefile` line 26 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/46#gh-1442312245:-Nn_nU11AbGLiuS35mLN:b-2cyh66) ([raw file](https://github.com/mkornacker/pixeltable/blob/e48c93ac64e2d97af6dd4c19d70a9b882ff2dd55/Makefile#L26)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

--with=test,docs
</blockquote></details>

merged test and docs into dev
___
*[`Makefile` line 51 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/46#gh-1442312008:-Nn_nXjjE7vpOhUsNeAO:b234cpx) ([raw file](https://github.com/mkornacker/pixeltable/blob/e48c93ac64e2d97af6dd4c19d70a9b882ff2dd55/Makefile#L51)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

what's that for?
</blockquote></details>

remove stray mp4 output from running tets
___
*[`.github/workflows/install_and_test.yml` line 1 at r3](https://reviewable.io/reviews/mkornacker/pixeltable/46#gh-1442310690:-Nn_nfOh7H7fFBfeb4VO:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/e48c93ac64e2d97af6dd4c19d70a9b882ff2dd55/.github/workflows/install_and_test.yml#L1)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Call this something like "Main CI"?
</blockquote></details>

Done.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,32 @@
+name: Run CI checks
```
Comment (mkornacker) [inline .github/workflows/ci.yml]: "CI checks"

**Review by mkornacker — APPROVED** [review]

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/46)* status: 0 of 14 files reviewed, 10 unresolved discussions (waiting on @mkornacker)

*[a discussion](https://reviewable.io/reviews/mkornacker/pixeltable/46#gh-1444079465:-NnfD1Y19B40oU2lCa2P:b-896fix) (no related file):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map [this GitHub inline comment thread](https://github.com/mkornacker/pixeltable/pull/46#discussion_r1444079465) to the right spot — sorry!)*

"CI checks"
</blockquote></details>

Done.

___

**Pull request merged**